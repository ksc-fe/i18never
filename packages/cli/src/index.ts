import { program } from 'commander';
import { tag } from './tag';
import { preCommit } from './preCommit';
import { initOptions } from '@i18never/shared';

export { preCommit, tag };

program.version(require('../package.json').version);

program
    .command('tag <path>')
    .option('-t, --token [token]', 'The token to request api.')
    .option('-s, --source [source]', 'Indicate where the keys are added from.')
    .option('-c, --config [configFile]', 'Specify the config file.')
    .description('Tag the appropriate file under the current file or folder.')
    .action((path, options) => {
        initOptions(options);
        tag(path);
    });

program
    .command('pre-commit')
    .option('-t, --token [token]', 'The token to request api.')
    .option('-s, --source [source]', 'Indicate where the keys are added from.')
    .option('-c, --config [configFile]', 'Specify the config file.')
    .description(
        'Check the temporary files and perform marking operations before git commit'
    )
    .action((options) => {
        initOptions(options);
        preCommit();
    });

program.parse(process.argv);
