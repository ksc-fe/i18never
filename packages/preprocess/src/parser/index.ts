// import { matchFileType } from '../helpers/utils';
import { parse as babelParse } from '@babel/parser';
import babelGenerate from '@babel/generator';
import traverse, { TraverseOptions } from '@babel/traverse';
import options from '../helpers/options';
// import parsePug from './parsePug';
// import parseJs from './parseJs';
// import parseVue from './parseVue';
// import transform from '../helpers/transform';
// import writeFile from '../helpers/writeFile';
import defaultVisitors, { KeyItem, Context } from '../visitors';

export enum FileType {
    JS = 'js',
    VUE = 'vue',
    PUG = 'pug',
    JSX = 'jsx',
    TSX = 'tsx',
    TS = 'ts',
}

// export async function parse(
//     source: string,
//     filename: string
// ): Promise<TempKeyItem[]> {
//     let keys: TempKeyItem[] = [];
//     let transResult: Array<TempKeyItem[]> = [];
//     if (!matchFileType(filename)) {
//         throw new Error('only Pug/Vue/Tsx/Jsx/Js syntax is supported');
//     }

//     const filesuffix = matchFileType(filename);

//     switch (filesuffix) {
//         case FileType.PUG:
//             keys = parsePug(source, filename);
//             break;
//         case FileType.VUE:
//             keys = parseVue(source, filename);
//             break;
//         case FileType.JSX:
//         case FileType.TSX:
//         case FileType.TS:
//         case FileType.JS:
//             keys = parseJs(source, filename, false, 1);
//             break;
//         default:
//             break;
//     }
//     transResult = await i18ntransform(keys, filename);
//     generateFile(filename, transResult);

//     return keys;
// }

export function generate(
    sourceCode: string,
    filename: string,
    visitors: TraverseOptions<Context> = defaultVisitors // for unit test
) {
    const keys: KeyItem[] = [];

    if (!options.matchChineseRE.test(sourceCode)) {
        return { keys, code: sourceCode };
    }

    const codeAst = babelParse(sourceCode, {
        sourceType: 'module',
        plugins: ['jsx', 'typescript'],
    });

    traverse(codeAst, visitors, undefined, {
        keys,
        filename,
    });
    const code = babelGenerate(codeAst, {
        jsescOption: { minimal: true },
        retainLines: true,
    }).code;

    return { keys, code };
}
