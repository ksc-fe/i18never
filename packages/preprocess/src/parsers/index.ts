import { KeyItem as BaseKeyItem } from '@i18never/shared';
import { Entity as JsEntity } from './js';
import { Entity as PugEntity } from './pug';
import { Entity as VueEntity } from './vue';
export { parse as jsParse } from './js';
export { parse as pugParse } from './pug';
export { parse as vueParse } from './vue';

export type KeyItem = BaseKeyItem<JsEntity | PugEntity | VueEntity>;
