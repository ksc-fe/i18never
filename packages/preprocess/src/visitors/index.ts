import { TraverseOptions } from '@babel/traverse';
import {
    ImportDeclaration,
    ObjectProperty,
    StringLiteral,
    TemplateLiteral,
    Context as BaseContext,
    KeyItem as BaseKeyItem,
} from '@i18never/shared';
import * as t from '@babel/types';
import { JSXText } from './JSXText';

type TextNode = t.StringLiteral | t.TemplateLiteral | t.JSXText;
export type KeyItem = BaseKeyItem<TextNode>;
export type Context = BaseContext<TextNode>;

export const visitors: TraverseOptions<Context> = {
    ImportDeclaration,
    ObjectProperty,
    StringLiteral,
    TemplateLiteral,
    JSXText,
};
