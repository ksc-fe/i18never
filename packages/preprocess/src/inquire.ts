import { getSdk, TagFragment, options } from '@i18never/shared';
import inquirer from 'inquirer';
import { KeyItem } from './parsers';

export type TranslationDetail = {
    language: string;
    tag: TagFragment;
};

export type InquireResultItem = {
    key: string;
    translationDetails: TranslationDetail[];
};

type Dict = Pick<KeyItem, 'key' | 'tags'> & {
    loc?: KeyItem['loc'];
};

export async function inquire(dicts: Dict[]) {
    const data = await queryTranslations(dicts);
    const result: InquireResultItem[] = [];

    for (const { key, tags } of dicts) {
        const translationDetails: TranslationDetail[] = [];
        const translations = data[key];

        if (!translations) {
            throw new Error(`Cannot find any translation for key: "${key}".`);
        }

        result.push({
            key,
            translationDetails,
        });

        const languages = Object.keys(translations);
        if (!tags) {
            for (const language of languages) {
                const tags = translations[language];

                // if the language's translation has multiple entries,
                // we should show inquirer for user to select the correct item
                if (tags.length > 1) {
                    await inquirer
                        .prompt<{ tag: TagFragment }>([
                            {
                                type: 'rawlist',
                                name: 'tag',
                                message: `Which translation do you need for string: "${key}" in '${language}' ?`,
                                choices: tags.map((tag) => {
                                    return {
                                        name: `Tag: "${tag.name}", Translation: "${tag.value}"`,
                                        value: tag,
                                    };
                                }),
                            },
                        ])
                        .then((answers) => {
                            translationDetails.push({
                                language,
                                tag: answers.tag,
                            });
                        });
                } else if (tags.length === 1) {
                    translationDetails.push({
                        language,
                        tag: tags[0],
                    });
                } else {
                    throw new Error(
                        `Can not find any translation for string: "${key}" [${language}].`
                    );
                }
            }
        } else {
            for (const language of languages) {
                const specifiedTag = tags[language] || 'default';
                const tag = translations[language].find(
                    (tag) => tag.name === specifiedTag
                );

                if (!tag) {
                    throw new Error(
                        `Can not find translation for string: "${key}" [${language}=${specifiedTag}]`
                    );
                }

                translationDetails.push({
                    language,
                    tag,
                });
            }
        }
    }

    return result;
}

async function queryTranslations(dicts: Dict[]) {
    const sdk = getSdk(options.uri);
    const { dicts: data } = await sdk.QueryOrCreateDicts({
        source: options.source,
        values: dicts.map(({ key, tags }) => {
            return {
                key,
                tags: tags
                    ? Object.keys(tags).map((language) => {
                          const name = tags[language];
                          return { language, name };
                      })
                    : undefined,
            };
        }),
    });

    return data.reduce((memo, item) => {
        memo[item.key] = item.translations.reduce((memo, item) => {
            memo[item.language] = item.tags;

            return memo;
        }, {} as Record<string, TagFragment[]>);

        return memo;
    }, {} as Record<string, Record<string, TagFragment[]>>);
}
