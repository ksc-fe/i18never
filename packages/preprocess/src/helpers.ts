import { TranslationDetail } from './inquire';
import { options, SupportExt } from '@i18never/shared';
import { jsParse, pugParse, vueParse } from './parsers';

export const extParserMap: {
    [key in SupportExt]: typeof jsParse | typeof pugParse | typeof vueParse;
} = {
    '.vue': vueParse,
    '.js': jsParse,
    '.ts': jsParse,
    '.mjs': jsParse,
    '.mts': jsParse,
    '.jsx': jsParse,
    '.tsx': jsParse,
    '.pug': pugParse,
    '.jade': pugParse,
};

export function getParserByExt(extname: string) {
    const parser = extParserMap[extname];
    if (!parser) {
        throw new Error(`${extname} file is not supported.`);
    }

    return parser;
}

export function getIdentifier(translationDetails: TranslationDetail[]) {
    const identifierTags: string[] = [];
    translationDetails.forEach(({ language, tag }) => {
        const tagName = tag.name;
        if (!tagName || tagName === 'default') {
            identifierTags.push(`${language}`);
        } else {
            identifierTags.push(`${language}=${tagName}`);
        }
    });

    return `${options.prefix}:${identifierTags.join(',')}`;
}
