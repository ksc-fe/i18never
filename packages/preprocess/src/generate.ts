import { KeyItem } from './parsers';
import { InquireResultItem } from './inquire';
import { getIdentifier } from './helpers';

export type ColumnOffset = [number, number];

export function generate(
    source: string,
    keys: KeyItem[],
    translations: InquireResultItem[]
) {
    const lines = source.split('\n');
    const lineColumnsOffset: ColumnOffset[][] = [];
    keys.forEach((item, index) => {
        const translationDetails = translations[index].translationDetails;
        const newIdentifier = getIdentifier(translationDetails);
        const oldIndentifer = item.identifier || '';
        if (newIdentifier !== oldIndentifer) {
            const loc = item.loc;
            const line = loc.line - 1; // line starts from 1

            let columnsOffset = lineColumnsOffset[line];
            if (!columnsOffset) {
                columnsOffset = lineColumnsOffset[line] = [];
            }

            lines[line] = updateSourceAndOffset(
                `[${newIdentifier}]`,
                oldIndentifer ? `[${oldIndentifer}]` : oldIndentifer,
                lines[line],
                columnsOffset,
                loc.column - 1
            );
        }
    });

    return lines.join('\n');
}

export function updateSourceAndOffset(
    newIdentifier: string,
    oldIndentifer: string,
    lineCode: string,
    columnsOffset: ColumnOffset[],
    column: number
) {
    let offset = 0;
    let index = 0;
    for (; index < columnsOffset.length; index++) {
        const [_column, _offset] = columnsOffset[index];
        if (_column >= column) break;
        offset += _offset;
    }

    const newColumn = column + offset;
    const oldLength = oldIndentifer.length;
    columnsOffset.splice(index, 0, [column, newIdentifier.length - oldLength]);

    return (
        lineCode.substring(0, newColumn) +
        newIdentifier +
        lineCode.substring(newColumn + oldLength)
    );
}
