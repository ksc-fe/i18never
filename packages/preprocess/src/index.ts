import { isIgnore, options } from '@i18never/shared';
import { jsParse, pugParse, KeyItem } from './parsers';
import fs from 'fs/promises';
import path from 'path';
import { inquire } from './inquire';
import { generate } from './generate';

export async function process(file: string) {
    const source = await fs.readFile(file, 'utf-8');
    const extname = path.extname(file);

    const keys = getKeys(source, extname);
    const translations = await inquire(keys);
    const newSource = generate(source, keys, translations);

    await fs.writeFile(newSource, file, 'utf-8');

    return newSource;
}

function getKeys(source: string, extname: string) {
    let keys: KeyItem[];
    switch (extname) {
        case '.js':
        case '.ts':
        case '.jsx':
        case '.tsx':
        case '.mjs':
            keys = jsParse(source);
            break;
        case '.pug':
            keys = pugParse(source);
            break;
        default:
            throw new Error(`${extname} file is not supported.`);
    }

    return keys.filter(({ identifier, key }) => {
        if (identifier) return !isIgnore(identifier);

        return options.matchRegexp.test(key);
    });
}
