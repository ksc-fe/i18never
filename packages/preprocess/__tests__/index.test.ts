import { process } from '../src';
import path from 'path';
import fs from 'fs/promises';
import { QueryOrCreateDictsQuery } from '@i18never/shared';

function MockGraphQLClient() {
    // emtpy
}
MockGraphQLClient.prototype.request =
    function (): Promise<QueryOrCreateDictsQuery> {
        return Promise.resolve(require('./assets/data.json'));
    };
jest.mock('graphql-request', () => {
    return {
        GraphQLClient: MockGraphQLClient,
    };
});

fs.writeFile = jest.fn().mockImplementation((source: string, file: string) => {
    // const extname = path.extname(file);
    // writeFileSync(
    //     `${file.slice(0, -extname.length)}.result${extname}`,
    //     source,
    //     'utf-8'
    // );
});

async function toMatchSnapshot(file: string) {
    expect(await process(path.resolve(__dirname, file))).toMatchSnapshot();
}

test('process js', async () => {
    await toMatchSnapshot('./assets/test.js');
});

test('process jsx', async () => {
    await toMatchSnapshot('./assets/react.jsx');
});

test('process tsx', async () => {
    await toMatchSnapshot('./assets/testTsx.tsx');
});

test('process pug', async () => {
    await toMatchSnapshot('./assets/testPug.pug');
});

test('process vue', async () => {
    await toMatchSnapshot('./assets/testVue.vue');
});

test('process pug in vue', async () => {
    await toMatchSnapshot('./assets/testPugInVue.vue');
});
