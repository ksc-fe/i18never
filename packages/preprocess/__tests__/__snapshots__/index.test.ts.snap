// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`process js 1`] = `
"export default class test {
    tabs = [
        {
            text: '[$_:en]测试',
            value: 'test',
        },
        {
            text: '[$_:en]测试一次',
            value: 'test one',
        },
    ];
}
"
`;

exports[`process jsx 1`] = `
"export default function Component() {
    const a = '[$_:en]测试';
    const b = '[$_:en]English';

    return () => {
        <div>
            <div>{a}</div>
            <div>{\`[$_:en,jp=v]购买\${a}台\`}</div>
            <div>[$_:en]购买</div>
            <Test prop=\\"[$_:en]属性\\" />
        </div>;
    };
}

function Test() {
    return <div>[$_:en]暂停任务</div>;
}
"
`;

exports[`process pug 1`] = `
"div
    | {{ aaa + '[$_:en]测试' + bbb + '[$_:en]购买' }}
    | {{ \`[$_:en,jp=v]购买\${a}台\` }}
    | [$_:en]购买{{ a }}[$_:en]测试
    Form(:hideOnSelect=\\"true\\")
        FormItem(
            :label=\\"istest ? '[$_:en]测试一次' : '[$_:en]购买'\\",
            :test=\\"\`[$_:en,jp=v]购买\${a}台\`\\",
            name=\\"[$_:en]测试\\") [$_:en]测试一次
    

"
`;

exports[`process pug in vue 1`] = `
"<script>
  let aa = '[$_:en]测试'
  let bb = '[$_:ignore]测试'
  let cc = \`[$_:en,jp=v]购买\${aa}台\`
</script>
<template lang=\\"pug\\">
div
  | {{ aaa + '[$_:en]测试' + bbb + '[$_:en]购买' }}
  | {{ \`[$_:en,jp=v]购买\${a}台\` }}
  Form(:hideOnSelect=\\"true\\")
      FormItem(
          :label=\\"istest ? '[$_:en]测试一次' : '[$_:en]购买'\\",
          :test=\\"\`[$_:en,jp=v]购买\${a}台\`\\",
          name=\\"[$_:en]测试\\") [$_:en]测试一次
</template>

"
`;

exports[`process tsx 1`] = `
"export default class test {
    test() {
        const a: { text: string; value: string }[] = [
            {
                text: '[$_:en]测试',
                value: 'test',
            },
        ];

        const b = <div id={\`[$_:en,jp=v]购买\${a}台\` as string}>[$_:en]购买</div>;

        return [a, b];
    }
}
"
`;

exports[`process vue 1`] = `
"<template>
    <div>
        <p>{{ aaa + '[$_:en]测试' + bbb + '[$_:en]购买' }}</p>
        <p>{{ \`[$_:en,jp=v]购买\${a}台\` }}</p>
        <p>[$_:en]测试</p>
        <Form>
            <FormItem
                :label=\\"istest ? '[$_:en]测试一次' : '[$_:en]购买'\\"
                :test=\\"\`[$_:en,jp=v]购买\${a}台\`\\"
            >
                <Input placeholder=\\"[$_:en]购买\\"></Input>
            </FormItem>
            <span>[$_:ignore]测试</span>
        </Form>
    </div>
</template>
<script>
export default {
    props: {
        value: {
            type: String,
            default: '[$_:en]测试',
        },
    },
    data() {
        return {
            test: '[$_:en]测试' + aaa + '[$_:en]测试' + bbb + '[$_:en]购买',
        };
    },
};
</script>
"
`;
